sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownUtils.scala

```diff
@@ -76,28 +78,48 @@ object PushDownUtils extends PredicateHelper {
    * @return the created `ScanConfig`(since column pruning is the last step of operator pushdown),
    *         and new output attributes after column pruning.
    */
-  // TODO: nested column pruning.
   def pruneColumns(
       scanBuilder: ScanBuilder,
       relation: DataSourceV2Relation,
-      exprs: Seq[Expression]): (Scan, Seq[AttributeReference]) = {
+      projects: Seq[NamedExpression],
+      filters: Seq[Expression]): (Scan, Seq[AttributeReference]) = {
     scanBuilder match {
+      case r: SupportsPushDownRequiredColumns if SQLConf.get.nestedSchemaPruningEnabled =>
+        val rootFields = SchemaPruning.identifyRootFields(projects, filters)
+        val prunedSchema = if (rootFields.nonEmpty) {
```

Comment:
there was a check in `prunePhysicalColumns`:

```
if (requestedRootFields.exists { root: RootField => !root.derivedFromAtt }) {
 ...
}
```

Is it removed from this move?

Reply:
This check was needed to detect if any nested column was requested. The old rule would not apply otherwise. Here, the situation is different as we need to prune top-level columns even if no nested attributes are requested.